example_external_provider_plugin/plugin.go

```diff
@@ -0,0 +1,215 @@
+	scopes := models.ScopeType{}
+
+	err := scopes.FromAwsAccountScope(models.AwsAccountScope{
+		Regions: &[]models.AwsRegion{
+			{
+				Name: "region1",
+				Vpcs: &[]models.AwsVPC{
+					{
+						Id: "vpc-1",
+						SecurityGroups: &[]models.AwsSecurityGroup{
+							{
+								Id: "sg-1",
+							},
+						},
+					},
+					{
+						Id: "vpc-2",
+						SecurityGroups: &[]models.AwsSecurityGroup{
+							{
+								Id: "sg-2",
+							},
+						},
+					},
+				},
+			},
+			{
+				Name: "region2",
+				Vpcs: &[]models.AwsVPC{
+					{
+						Id: "vpc3",
+						SecurityGroups: &[]models.AwsSecurityGroup{
+							{
+								Id: "sg3",
+							},
+						},
+					},
+				},
+			},
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	scopesB, err := scopes.MarshalJSON()
+	if err != nil {
+		return nil, err
+	}
+
+	return &provider_service.DiscoverScopesResult{
+		Scopes: string(scopesB),
+	}, nil
+}
+
+func (p *Provider) DiscoverInstances(context.Context, *provider_service.DiscoverInstancesParams) (*provider_service.DiscoverInstancesResult, error) {
+	// Here needs to discover all instances according to the ScanScope parameter.
+	// ScanScope can be added to the backend api, and can contain fields like instance tags, location etc.
+	return &provider_service.DiscoverInstancesResult{
+		Instances: []*provider_service.Instance{
+			{
+				Id:           "instanceID1",
+				Location:     "region1",
+				Image:        "ami-01d08089481510ba2",
+				InstanceType: "t2.large",
+				Platform:     "Linux/UNIX",
+				Tags: []*provider_service.Tag{
+					{
+						Key: "Name",
+						Val: "instance1",
+					},
+				},
+				LaunchTime: timestamppb.New(time.Now()),
+			},
+			{
+				Id:           "instanceID2",
+				Location:     "region2",
+				Image:        "ami-01d08089481510ba2",
+				InstanceType: "t2.large",
+				Platform:     "Linux/UNIX",
+				Tags: []*provider_service.Tag{
+					{
+						Key: "Name",
+						Val: "instance2",
+					},
+				},
+				LaunchTime: timestamppb.New(time.Now()),
+			},
+		},
+	}, nil
+}
+
+func (p *Provider) GetInstanceRootVolume(context.Context, *provider_service.GetInstanceRootVolumeParams) (*provider_service.GetInstanceRootVolumeResult, error) {
+	return &provider_service.GetInstanceRootVolumeResult{
+		Volume: &provider_service.Volume{
+			Id:       "volume1",
+			Location: "region1",
+		},
+	}, nil
+}
+
+func (p *Provider) WaitForInstanceReady(context.Context, *provider_service.WaitForInstanceReadyParams) (*provider_service.WaitForInstanceReadyResult, error) {
+	return &provider_service.WaitForInstanceReadyResult{}, nil
+}
+
+func (p *Provider) DeleteInstance(context.Context, *provider_service.DeleteInstanceParams) (*provider_service.DeleteInstanceResult, error) {
+	return &provider_service.DeleteInstanceResult{}, nil
+}
+
+func (p *Provider) AttachVolumeToInstance(context.Context, *provider_service.AttachVolumeToInstanceParams) (*provider_service.AttachVolumeToInstanceResult, error) {
```

Comment:
Here the snapshot volume needs to be attached to the running scanning vm (after WaitForInstanceReady). VMClarity will verify that it is attached be calling WaitForVolumeAttached.

I would move the methods move close to one another so the relation will be more clear (e.g. put WaitForVolumeAttached after this one)